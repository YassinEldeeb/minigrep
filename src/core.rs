use colored::Colorize;

#[derive(PartialEq, Eq, Debug)]
pub struct Match<'a> {
    index: usize,
    line: &'a str,
}

impl<'a> Match<'a> {
    fn new(line: &'a str, index: usize) -> Self {
        Match { line, index }
    }
}

pub fn search<'a>(query: &str, content: &'a str) -> Vec<Match<'a>> {
    let mut matches: Vec<Match> = vec![];

    for (idx, line) in content.lines().enumerate() {
        if line.contains(query) {
            matches.push(Match::new(line, idx));
        }
    }

    matches
}

pub fn colorize_matches(matches: Vec<Match>, query: &str) -> Vec<String> {
    let mut colorized: Vec<String> = Vec::new();

    for e in matches {
        let colored = str::replace(e.line, query, &query.bright_purple().bold().to_string());

        let formatted = format!("{}. {}", (e.index + 1).to_string(), colored);
        colorized.push(formatted);
    }

    colorized
}

#[cfg(test)]
mod tests {
    use super::*;

    const QUERY: &str = "programming";
    const CONTENT: &str = "\
The most loved programming languages are:
1. Typescript.
2. The Rust Programming Langauge.
3. Kotlin.
        ";

    #[test]
    fn search_case_insensitive() {
        assert_eq!(
            search(QUERY, CONTENT),
            vec![
                Match::new("The most loved programming languages are:", 0),
                Match::new("2. The Rust Programming Langauge.", 2)
            ]
        );
    }

    #[test]
    fn search_case_sensitive() {
        assert_eq!(
            search(QUERY, CONTENT),
            vec![Match::new("The most loved programming languages are:", 0)]
        );
    }

    #[test]
    fn gets_nothing_when_no_matches() {
        let query = "Rust";
        let content = "How is it going?";

        assert_eq!(search(query, content), vec![] as Vec<Match>);
    }
}
