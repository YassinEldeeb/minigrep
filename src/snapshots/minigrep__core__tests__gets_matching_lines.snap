---
source: src/core.rs
assertion_line: 38
expression: "search(query, content)"

---
[
    "3. 2. the \u{1b}[1;95mRust\u{1b}[0m programming langauge.",
]
